homeassistant/components/habitica/calendar.py

```diff
@@ -0,0 +1,179 @@
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        return [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.TODO
+            and not task["completed"]
+            and task.get("date")
+            and (start := to_date(task["date"]))
+            and (start_date.date() <= start <= end_date.date())
+        ]
+
+
+class HabiticaDailiesCalendarEntity(HabiticaCalendarEntity):
+    """Habitica dailies calendar entity."""
+
+    entity_description = CalendarEntityDescription(
+        key=HabiticaCalendar.DAILIES,
+        translation_key=HabiticaCalendar.DAILIES,
+    )
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the next upcoming event."""
+
+        today = datetime.combine(
+            dt_util.now().date(), time(), dt_util.DEFAULT_TIME_ZONE
+        )
+        events = [
+            CalendarEvent(
+                start=(start := next_recurrence.date()),
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.DAILY
+            and not task["completed"]
+            and task["everyX"]
+            if (next_recurrence := build_rrule(task).after(today, inc=True))
+        ]
+        events_sorted = sorted(
+            events,
+            key=lambda event: (
+                event.start,
+                self.coordinator.data.user["tasksOrder"]["dailys"].index(event.uid),
+            ),
+        )
+
+        return events_sorted[0] if events_sorted else None
+
+    async def async_get_events(
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Return calendar events within a datetime range."""
+
+        today = to_date(self.coordinator.data.user["lastCron"])
+        # returns only todays and future dailies.
+        # If a daily is completed it will not be shown for today but still future recurrences
+        # If the cron hasn't run, not completed dailies are yesterdailies and displayed yesterday
+        return [
+            CalendarEvent(
+                start=start,
+                end=start + timedelta(days=1),
+                summary=task["text"],
+                description=task["notes"],
+                uid=task["id"],
+                rrule=str(build_rrule(task))[30:],
+            )
+            for task in self.coordinator.data.tasks
+            if task["type"] == HabiticaTaskType.DAILY and task["everyX"]
```

Comment:
yes, there are dailies without a recurrence, they are called grey dailies and they can be completed whenever but become never due.
rrule seems to not be able to parse that.

Reply:
Does that mean they show up every day? or more? If so then maybe that can be set in the rrule

Reply:
on the dailies list in habitica all tasks show up, by default, although one can filter for dailies that are due or grey(completed or not due today). It is still possible to complete dailies that are not due to get the reward but they won't cause damage if they are not completed, so technically grey tasks are optional. An task without a recurrence is always optional, so I think they shouldn't show up in the calendar at all.

Reply:
I must correct myself, `rrule` is not to blame, it can parse a 0 interval correctly, it's Home Assistant that freaks out, i think it's getting into an infinite loop somewhere

Reply:
I made a bug report for that ^^
- https://github.com/home-assistant/core/issues/128396